src/compiler/program.ts

```diff
@@ -194,7 +197,117 @@ namespace ts {
             getEnvironmentVariable: name => system.getEnvironmentVariable ? system.getEnvironmentVariable(name) : "",
             getDirectories: (path: string) => system.getDirectories(path),
             realpath,
-            readDirectory: (path, extensions, include, exclude, depth) => system.readDirectory(path, extensions, include, exclude, depth)
+            readDirectory: (path, extensions, include, exclude, depth) => system.readDirectory(path, extensions, include, exclude, depth),
+            createDirectory: d => system.createDirectory(d)
+        };
+        return compilerHost;
+    }
+
+    /*@internal*/
+    export function changeCompilerHostToUseCache(
+        host: CompilerHost,
+        toPath: (fileName: string) => Path,
+        useCacheForSourceFile: boolean
+    ) {
+        const originalReadFile = host.readFile;
+        const originalFileExists = host.fileExists;
+        const originalDirectoryExists = host.directoryExists;
+        const originalCreateDirectory = host.createDirectory;
+        const originalWriteFile = host.writeFile;
+        const originalGetSourceFile = host.getSourceFile;
+        const readFileCache = createMap<string | false>();
+        const fileExistsCache = createMap<boolean>();
+        const directoryExistsCache = createMap<boolean>();
+        const sourceFileCache = createMap<SourceFile>();
+
+        const readFileWithCache = (fileName: string): string | undefined => {
+            const key = toPath(fileName);
+            const value = readFileCache.get(key);
+            if (value !== undefined) return value || undefined;
+            return setReadFileCache(key, fileName);
+        };
+        const setReadFileCache = (key: Path, fileName: string) => {
+            const newValue = originalReadFile.call(host, fileName);
+            readFileCache.set(key, newValue || false);
+            return newValue;
+        };
+        host.readFile = fileName => {

```

Comment:
I agree with Nathan - we shouldn't be mutating the host object. Since we can recieve these from API consumers what happens here if, for example, this member is marked non-configurable? It's much safer to build a cacheing host that wraps the original host.

Reply:
Agreed but the API is internal because we haven't fixed it yet and hence this temporary. This will move to creating host for builder once we are there. And hence TODO.

Reply:
Mmmm we're going to want an issue open to track fixing this particular bit of tech debt, then. Otherwise we'll probably forget it until we accidentally break someone by using it somewhere in, eg, the language service. It can get away with mutating right now because tsc and tsbuild don't have any API hooks yet; but once tsbuild has a public API or we use this in a layer which has a public API, it'll need to be fixed.

Reply:
Created #28640 to track this